jwks.go

```diff
@@ -180,22 +191,84 @@ func (j *JWKS) RawJWKS() []byte {
-		keys[kid] = cryptoKey.public
+	for _, key := range j.keys {
+		keys[key.kid] = key.Public
 	}
 	j.mux.Unlock()
 	return keys
 }
+func findKeyIdx(alg string, kid string, keys []ParsedJWK) int {
+	for idx, key := range keys {
+		if key.kid == kid {
+			return idx
+		}
+	}
+	return -1
+}
+
+func (j *JWKS) canUseKey(key ParsedJWK) bool {
+	canUseKey := true
+	// jwkUseWhitelist might be empty if the jwks was from keyfunc.NewJSON() or if JWKUseNoWhitelist option was true.
+	// in this case we don't restrict "use" parameter
+	if len(j.jwkUseWhitelist) > 0 {
+		_, canUseKey = j.jwkUseWhitelist[key.use]
+	}
+	return canUseKey
+}
+
+func (j *JWKS) getMatchingKeys(alg, kid, iss string) []*ParsedJWK {
+	result := make([]*ParsedJWK, 0, 1)
+	if kid != "" {
+		for idx, key := range j.keys {
+			if (key.kid == kid) && (key.algorithm == alg || key.algorithm == "") && j.canUseKey(key) {
+					result = append(result, &j.keys[idx])
+			}
+		}
+		return result
+	}
+	if iss != "" {
+		for _, key := range j.keys {
+			if key.kid == iss && (key.algorithm == alg || key.algorithm == "") && j.canUseKey(key) {
+				result = append(result, &key)
+			}
+		}
+	}
+	// no match with "iss", use "alg" only
+	if len(result) == 0 {
+		for idx, key := range j.keys {
+			if key.algorithm == alg || key.algorithm == "" {
+				// jwkUseWhitelist might be empty if the jwks was from keyfunc.NewJSON() or if JWKUseNoWhitelist option was true.
+				canUseKey := true
+				if len(j.jwkUseWhitelist) > 0 {
+					_, canUseKey = j.jwkUseWhitelist[j.keys[idx].use]
+				}
+				if canUseKey {
+					result = append(result, &j.keys[idx])
+				}
+			}
+		}
+	}
+	return result
+}
 
-// getKey gets the jsonWebKey from the given KID from the JWKS. It may refresh the JWKS if configured to.
-func (j *JWKS) getKey(alg, kid string) (jsonKey interface{}, err error) {
+// GetMatchingKeysWithRefresh implements the logic described in
+// https://docs.singlestore.com/db/v7.8/en/security/authentication/authenticate-via-jwt.html
+//
+// JWTs are matched with JSON Web Keys (JWKs) for validation as follows:
+//
+// 1. If the JWT has a kid (Key ID) field, the JWKs with matching kid fields are validated.
+// 2. If the JWT has a kid field that doesn’t match any JWK or jwt_config key, the authentication request is rejected. See Validate JWTs with the jwt-config for more information.
+// 3. If the JWT has an iss (Issuer) field (instead of a kid field) that matches the kid in one or more JWKs, the JWKs with matching kid fields are validated.
+// 4. If the JWT does not have a kid field and the iss field does not match the kid field in any JWK, then validation is attempted with all the JWKs with a matching alg (Algorithm) field. If the alg field is not specified, the kty (Key Type) field is used instead.
+//
+func (j *JWKS) GetMatchingKeysWithRefresh(alg, kid, iss string) (matchingKeys []*ParsedJWK, err error) {
 	j.mux.RLock()
-	pubKey, ok := j.keys[kid]
+	matchingKeys = j.getMatchingKeys(alg, kid, iss)
```

Comment:
```suggestion
        func() {
             j.mux.RLock()
             defer j.mux.RUnlock()
	     matchingKeys = j.getMatchingKeys(alg, kid, iss)
	}()
```

Unless it's really hard, unlock should be done with `defer` so that an un-expected panic doesn't lead to program hangs

Reply:
makes sense, I'm moving lock and unlock to `getMatchingKeys` itself as it's always invoked with it.